### ISSUE TITLE
documentation for unconstrain_pars, constrain_pars is insufficient and misleading

### ISSUE THREAD

**Issue by jgabry** [stan-dev/rstan#344]
#### Summary:

Currently the documentation for the `unconstrain_pars` and `constrain_pars` functions is lacking and even misleading
#### Description:
- there are no usage examples
- the argument descriptions do not sufficiently describe the form of the required inputs
- there are some important subtleties that are not described in the documentation and that can lead to incorrect usage. For example, `unconstrain_pars(object, pars)`, returns the same exact thing even if the parameter values provided in `pars` are permuted. This is particularly confusing  because the values are unnamed and it is not documented at all.
#### Reproducible Steps:

Here's an example of the last bullet point above: 

``` r
fit <- stan_demo("eight_schools")
pars <- list(mu = 0, tau = 2, theta = rep(1.5, 8))
unconstrain_pars(fit, pars)
unconstrain_pars(fit, rev(pars))
```

Those two calls to `unconstrain_pars` return the same thing even though the parameters are passed in the opposite order. The both return an unnamed vector:

```
 [1] 0.0000000 1.5000000 1.5000000 1.5000000 1.5000000 1.5000000
 [7] 1.5000000 1.5000000 1.5000000 0.6931472
```

This seems like a recipe for a lot of confusion. At a minimum we need to document this (I can do that), but I wonder if we should also be naming the values returned by `unconstrain_pars`? Why are the values returned by `constrain_pars` named but not those returned by `unconstrain_pars`? 
#### RStan Version:

2.12.1

**Comment by bgoodri** [comment]
The unconstrained parameters do not have names and their sizes do not always correspond 1-to-1 with the constrained parameters. For example, a simplex is longer and a unit vector is shorter. I think we will have to overhaul this if we go to reference classes.

**Comment by jgabry** [comment]
Yeah that makes sense about the sizes. I'll try to add a few things to the doc to make all of this clear, in particular the bit about the order of the parameters in the returned object being invariant to the order in which they are specified. It seems strange to allow them to be specified in any order if that order isn't respected in the returned object, especially if we don't say this anywhere.

**Comment by maverickg** [comment]
In your example, it's a named list for the second argument.  It's order
should not matter.

**Comment by jgabry** [comment]
@maverickg That's true, good point. I'm more concerned that there's no indication anywhere what the order of the returned object will be. Also a few other things related to `constrain_pars`/`unconstrain_pars` that I find strange:
- If you pass an unnamed list to `unconstrain_pars` it gives an unhelpful error message (actually not only unhelpful but misleading):

``` r
> unconstrain_pars(fit, unname(pars))
Error in eval(substitute(expr), envir, enclos) : 
  expecting a string vector
Called from: object@.MISC$stan_fit_instance$unconstrain_pars(pars)
```

which I know refers to needing the names for the list but from the error message one could easily conclude that `pars` should be a character vector instead of a named list. The documentation doesn't actually say the list needs to be named (it just says "An [sic] list specifying the values for all parameters on the constrained space"), so we either need to say that it must be named or provide an error message that's helpful. Preferably both. Currently we are providing very little guidance to users here. The arguments aren't described clearly and there are zero examples. We need to either have very clear argument documentation or plenty of examples (both really) but we've never had either of those for these functions.
- For `constrain_pars`, the vector `upars` can have arbitrary names

``` r
pars <- list(mu = 0, tau = 2, theta = rep(1.5, 8))
up <- unconstrain_pars(fit, pars)
names(up) <- letters[seq_along(up)]
constrain_pars(fit, up)
```

but I think that should lead to at least a warning.

**Comment by bob-carpenter** [comment]
You can reference the manual for precise definitions of the transforms.  I think they're all in the right orders, but if not, it'd be good to know.

**Comment by bgoodri** [comment]
I think they are fine on the Stan side. But in the future, rstan needs to
expose the transformations so that we can transform one constrained
parameter at a time.

**Comment by jgabry** [comment]
Yeah I don't think the order is wrong it's just not clear to the user what
the order actually is without extra research. And we don't
currently indicate in the rstan doc where they can find that information.

Being able to transform one at a time would be fantastic. That would be
useful for a variety of things and it would also make it much easier to
document things like this for the user.

**Comment by davharris** [comment]
This issue is closely related to the documentation for `log_prob`, which I tried to address in PR #347.

**Comment by jgabry** [comment]
Hey David, thanks for following up. We're a bit swamped at the moment but
we haven't forgotten about this and your other PR. We'll try to deal with
everything all at once before the next release.

**Comment by jan-glx** [comment]
Landed here searching for names of  the unconstrained parameters returned by `unconstrain_pars`.
`your_stan_fit@.MISC$stan_fit_instance$unconstrained_param_names(FALSE, FALSE)` is an answer I found in the source, in case you come here for the same reason... Sorry for the noise, everybody else.
